node/lib/util/cherry_pick_util.js

```diff
@@ -0,0 +1,569 @@
+            0 === Object.keys(picks.commits).length) {
+            yield CherryPickFileUtil.cleanCherryPick(metaRepo.path());
+            throw new UserError("Nothing to commit.");
+        }
+        result.newMetaCommit = yield finish(metaRepo, metaIndex, commit);
+    }
+    return result;
+});
+
+/**
+ * Continue the in-progress cherry-pick in the specified `repo`.  Throw a
+ * `UserError` if the continue cannot be initiated, e.g., because there is not
+ * a cherry-pick in progress or there are still conflicts.  Return an object
+ * describing the commits that were made and any errors that were generated.
+ *
+ * @async
+ * @param {NodeGit.Repository} repo
+ * @return {Object}      return
+ * @return {String|null} return.newMetaCommit
+ * @return {Object}      returm.submoduleCommits
+ * @return {Object}      returm.newSubmoduleCommits
+ * @return {String|null} return.errorMessage
+ */
+exports.continue = co.wrap(function *(repo) {
+    assert.instanceOf(repo, NodeGit.Repository);
+
+    const status = yield StatusUtil.getRepoStatus(repo);
+    const cherryPick = status.cherryPick;
+    if (null === cherryPick) {
+        throw new UserError("No cherry-pick in progress.");
+    }
+    if (status.isConflicted()) {
+        throw new UserError("Resolve conflicts then continue cherry-pick.");
+    }
+    const index = yield repo.index();
+    const commit = yield repo.getCommit(cherryPick.picked);
+
+    const subResult = yield RebaseUtil.continueSubmodules(repo,
+                                                          index,
+                                                          status,
+                                                          commit);
+    yield index.write();
+    const result = {
+        newMetaCommit: null,
+        submoduleCommits: subResult.commits,
+        newSubmoduleCommits: subResult.newCommits,
+        errorMessage: subResult.errorMessage,
+    };
+    if (null === subResult.errorMessage) {
+        if (status.isIndexDeepClean() &&
+            0 === Object.keys(subResult.commits).length &&
+            0 === Object.keys(subResult.newCommits).length) {
+            yield CherryPickFileUtil.cleanCherryPick(repo.path());
+            throw new UserError("Nothing to commit.");
+        }
+
+        result.newMetaCommit = yield finish(repo, index, commit);
+    }
+    return result;
+});
+
+/**
+ * Abort the cherry-pick in progress in the specified `repo` and return the
+ * repository to exactly the state of the initial commit.  Throw a `UserError`
+ * if no cherry-pick is in progress.
+ *
+ * @param {NodeGit.Repository} repo
+ */
+exports.abort = co.wrap(function *(repo) {
+    assert.instanceOf(repo, NodeGit.Repository);
+
+    const cherry = yield CherryPickFileUtil.readCherryPick(repo.path());
+    if (null === cherry) {
+        throw new UserError("No cherry-pick in progress.");
+    }
+    const commit = yield repo.getCommit(cherry.originalHead);
+    yield Reset.reset(repo, commit, Reset.TYPE.HARD);
```

Comment:
I think we can start a cherry-pick in an unclean state, right?  That is, this reset might blow away local changes Also, let's say you start a cherry-pick, then get distracted and start making changes in some other submodule.  It seems like we shouldn't delete those changes?

Reply:
No, we will not let you start a pick in an unclean state. Partially because of this issue, and partially because it's extremely difficult to do so (we'd need some sort of dry-run that libgit2 rebase doesn't provide). I have a TODO about allowing it to run in an unclean state on line 421 of `util/cherry_pick_util.js`.

Now, re. files that may have been modified during the cherry-pick for whatever reason, I have a TODO on line 68 of `cmd/cherry_pick.js` where I tell the users that the command blows away all their changes.

Reply:
I somehow missed the code that prevents starting in an unclean state -- can you point me to it?

Reply:
from line 420 of `util/cherry_pick_util.js`

```JavaScript
    if (!status.isDeepClean(false)) {
        // TODO: Git will refuse to run if there are staged changes, but will
        // attempt a cherry-pick if there are just workdir changes.  We should
        // support this in the future, but it basically requires us to dry-run
        // the rebases in all the submodules, and I'm uncertain how to do that
        // at the moment.

        throw new UserError(`\
The repository has uncommitted changes.  Please stash or commit them before
running cherry-pick.`);
    }
```

Reply:
Oh, I even commented on it. OK, good.